client/landing/stepper/hooks/use-site-copy.tsx

```diff
@@ -0,0 +1,80 @@
+import { WPCOM_FEATURES_COPY_SITE } from '@automattic/calypso-products';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useDispatch } from '@wordpress/data';
+import { useEffect, useMemo, useCallback } from 'react';
+import { useDispatch as useReduxDispatch, useSelector } from 'react-redux';
+import { useQuerySitePurchases } from 'calypso/components/data/query-site-purchases';
+import { ONBOARD_STORE } from 'calypso/landing/stepper/stores';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { clearSignupDestinationCookie } from 'calypso/signup/storageUtils';
+import { getCurrentUserId } from 'calypso/state/current-user/selectors';
+import {
+	hasLoadedSitePurchasesFromServer,
+	isFetchingSitePurchases,
+} from 'calypso/state/purchases/selectors';
+import siteHasFeature from 'calypso/state/selectors/site-has-feature';
+import { fetchSiteFeatures } from 'calypso/state/sites/features/actions';
+import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+
+function useSafeSiteHasFeature( siteId: number, feature: string ) {
+	const dispatch = useReduxDispatch();
+	useEffect( () => {
+		dispatch( fetchSiteFeatures( siteId ) );
+	}, [ dispatch, siteId ] );
+
+	return useSelector( ( state ) => {
+		return siteHasFeature( state, siteId, feature );
+	} );
+}
+
+export const useSiteCopy = ( site: SiteExcerptData ) => {
+	const userId = useSelector( ( state ) => getCurrentUserId( state ) );
+	const hasCopySiteFeature = useSafeSiteHasFeature( site.ID, WPCOM_FEATURES_COPY_SITE );
+	const plan = site?.plan;
+	const isSiteOwner = site.site_owner === userId;
+	useQuerySitePurchases( site.ID );
+	const isLoadingPurchase = useSelector(
+		( state ) => isFetchingSitePurchases( state ) || ! hasLoadedSitePurchasesFromServer( state )
+	);
+
+	const { setPlanCartItem } = useDispatch( ONBOARD_STORE );
+
+	const shouldShowSiteCopyItem = useMemo( () => {
+		return hasCopySiteFeature && isSiteOwner && plan && ! isLoadingPurchase;
+	}, [ hasCopySiteFeature, isSiteOwner, plan, isLoadingPurchase ] );
+
+	const startSiteCopy = useCallback(
+		//eslint-disable-next-line @typescript-eslint/no-explicit-any
+		( eventName: string, extraProps?: Record< string, any > ) => {
+			if ( ! plan ) {
+				return;
+			}
+			clearSignupDestinationCookie();
+			setPlanCartItem( { product_slug: plan.product_slug } );
+			recordTracksEvent( eventName, extraProps );
+		},
+		[ plan, setPlanCartItem ]
+	);
+
+	return useMemo(
+		() => ( {
+			shouldShowSiteCopyItem,
+			startSiteCopy,
+		} ),
+		[ shouldShowSiteCopyItem, startSiteCopy ]
+	);
+};
+
+export const withSiteCopy = createHigherOrderComponent(
```

Comment:
Do we need a higher order component because we're using Calypso store data?

Reply:
we need a HOC because we use a hook.

Reply:
> we need a HOC because we use a hook.

@kozer This hook doesn't seem to need a HOC though?

https://github.com/Automattic/wp-calypso/blob/f31905ccef603a0f15c963177574022ffd7ed915/client/data/sites/use-site-excerpts-query.ts#L27-L40

What's the difference between the two?

Reply:
It doesn't seem that we use it inside a class component. Do we? The moment we need to do that, we will need to wrap it inside a functional component (which is what the HOC I created does.). Hooks can't be used inside class components.

Reply:
> Hooks can't be used inside class components.

@kozer Ah, this makes sense. Thanks for satisfying my curiosity!